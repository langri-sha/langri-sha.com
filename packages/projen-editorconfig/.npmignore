*.test.*
.*
__snapshots__/
tsconfig.json
