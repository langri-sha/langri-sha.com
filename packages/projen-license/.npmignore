# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
*.test.*
.*
__snapshots__/
tsconfig*.json
