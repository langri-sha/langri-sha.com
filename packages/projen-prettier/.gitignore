# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/package.json
!/tsconfig.json
!/license
!/.npmignore
!/tsconfig.build.json
!/.projenrc.ts
