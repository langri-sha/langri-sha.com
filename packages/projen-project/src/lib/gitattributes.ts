import {
  GitAttributesFile as BaseGitAttributesFile,
  type IResolver,
} from 'projen'

export class GitAttributesFile extends BaseGitAttributesFile {
  protected override synthesizeContent(
    resolver: IResolver,
  ): string | undefined {
    const content = super.synthesizeContent(resolver)?.split('\n')

    if (!content) {
      return
    }

    const [marker, blank, ...rest] = content
    rest.sort((a, b) => a.localeCompare(b))

    return [marker, blank, ...rest, ''].join('\n')
  }
}
