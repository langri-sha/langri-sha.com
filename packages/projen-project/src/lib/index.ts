export * from './node-package.js'
export * from './projenrc.js'
