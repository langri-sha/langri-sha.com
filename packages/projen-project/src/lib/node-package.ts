import * as fs from 'node:fs'

import { Project, javascript } from 'projen'

/**
 * Options for configuring Node.js packages.
 */
export interface NodePackageOptions extends javascript.NodePackageOptions {
  peerDependenciesMeta?: Record<string, { optional?: boolean }>
  type?: 'commonjs' | 'module'
}

/**
 * Custom Node.js package manager that preserves package.
 */
export class NodePackage extends javascript.NodePackage {
  constructor(project: Project, options?: NodePackageOptions) {
    super(project, options)

    this.addVersion(this.#getPackageJson()?.version ?? '0.0.0')

    if (options?.type) {
      this.addField('type', options.type)
    }

    if (options?.peerDependenciesMeta) {
      this.addField('peerDependenciesMeta', options.peerDependenciesMeta)
    }

    // Reset, so we can run `pnpx projen` on CI after Renovate upgrades.
    this.project.tasks
      .tryFind('install:ci')
      ?.reset(this.project.tasks.tryFind('install')?.steps[0]?.exec)
  }

  #getPackageJson() {
    try {
      return JSON.parse(fs.readFileSync(this.file.absolutePath, 'utf-8'))
    } catch (_) {
      return undefined
    }
  }

  override postSynthesize(): void {
    if (this.project.parent) {
      return
    }

    super.postSynthesize()
  }
}
