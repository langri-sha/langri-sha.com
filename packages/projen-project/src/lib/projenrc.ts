import { ProjenrcFile as BaseProjenrcFile, Project, SampleFile } from 'projen'

export interface ProjenrcOptions {
  /**
   * The name of the projenrc file.
   * @default ".projenrc.mts"
   */
  readonly filename?: string
}

export class ProjenrcFile extends BaseProjenrcFile {
  public override readonly filePath: string

  constructor(
    project: Project,
    { filename = '.projenrc.mts' }: ProjenrcOptions,
  ) {
    super(project)

    this.filePath = filename
    this.project.defaultTask?.exec(
      `node --loader ts-node/esm/transpile-only ${filename}`,
    )

    new SampleFile(project, filename, {
      contents: `import { Project } from '@langri-sha/projen-project'

      const project = new Project()

      project.synth()
      `,
    })
  }
}
