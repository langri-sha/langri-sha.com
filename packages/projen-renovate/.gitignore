# ~~ Generated by projen. To modify, edit .projenrc.mts and run "npx projen".
renovate.d.ts
