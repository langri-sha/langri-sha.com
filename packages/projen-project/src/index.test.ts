import { synthSnapshot } from 'projen/lib/util/synth'
import { describe, expect, test } from '@langri-sha/jest-test'

import { Project } from './index'
import { Husky } from '@langri-sha/projen-husky'
import { EditorConfig } from '@langri-sha/projen-editorconfig'
import { Renovate } from '@langri-sha/projen-renovate'
import { Codeowners } from '@langri-sha/projen-codeowners'
import { Beachball } from '@langri-sha/projen-beachball'
import { TypeScriptConfig } from '@langri-sha/projen-typescript-config'
import * as path from 'node:path'
import { License } from '@langri-sha/projen-license'
import { NodePackage, ProjenrcFile } from './lib'
import { IgnoreFile } from 'projen'
import { JestConfig } from '@langri-sha/projen-jest-config'
import { Prettier } from '@langri-sha/projen-prettier'
import { ESLint } from '@langri-sha/projen-eslint'
import { LintStaged } from '@langri-sha/projen-lint-staged'

test('defaults', () => {
  const project = new Project({
    name: 'test-project',
  })

  expect(synthSnapshot(project)).toMatchSnapshot()
  expect(project.beachball).toBeUndefined()
  expect(project.codeowners).toBeUndefined()
  expect(project.editorConfig).toBeUndefined()
  expect(project.eslint).toBeUndefined()
  expect(project.husky).toBeUndefined()
  expect(project.jestConfig).toBeUndefined()
  expect(project.license).toBeUndefined()
  expect(project.lintStaged).toBeUndefined()
  expect(project.npmIgnore).toBeUndefined()
  expect(project.package).toBeUndefined()
  expect(project.prettier).toBeUndefined()
  expect(project.projenrc).toBeInstanceOf(ProjenrcFile)
  expect(project.renovate).toBeUndefined()
  expect(project.typeScriptConfig).toBeUndefined()
})

test('add subproject', () => {
  const project = new Project({
    name: 'test-project',
  })

  const sub = project.addSubproject({
    name: '@someproject/test',
    outdir: path.join('someproject', '@some', 'test'),
    typeScriptConfigOptions: {},
  })

  expect(synthSnapshot(project)).toMatchSnapshot()
  expect(sub.projenrc).toBeInstanceOf(ProjenrcFile)
})

test('find subproject', () => {
  const project = new Project({
    name: 'test-project',
  })

  const subproject = project.addSubproject({
    name: '@someproject/test',
    outdir: path.join('sub', '@some', 'test'),
    typeScriptConfigOptions: {},
  })

  subproject.addSubproject({
    name: '@someproject/test2',
    outdir: path.join('subsub', '@some', 'test2'),
    typeScriptConfigOptions: {},
  })

  expect(project.findSubproject('@someproject/test')).toBeInstanceOf(Project)
  expect(project.findSubproject('@someproject/test2')).toBeInstanceOf(Project)
  expect(() => project.findSubproject('non-existing')).toThrowError(
    'Cannot find subproject non-existing',
  )
})

test('with Beachball configuration', () => {
  const project = new Project({
    name: 'test-project',
    package: {},
    beachballOptions: {},
    prettierOptions: {},
    typeScriptConfigOptions: {},
  })

  expect(synthSnapshot(project)).toMatchSnapshot()
  expect(project.beachball).toBeInstanceOf(Beachball)
})

test('with code owners configured', () => {
  const project = new Project({
    name: 'test-project',
    codeownersOptions: {
      '*': '@admin',
    },
  })

  expect(synthSnapshot(project)).toMatchSnapshot()
  expect(project.codeowners).toBeInstanceOf(Codeowners)
})

test('with EditorConfig options', () => {
  const project = new Project({
    name: 'test-project',
    editorConfigOptions: {},
    prettierOptions: {},
  })

  expect(synthSnapshot(project)).toMatchSnapshot()
  expect(project.editorConfig).toBeInstanceOf(EditorConfig)
})

test('with ESLint options', () => {
  const project = new Project({
    name: 'test-project',
    package: {},
    eslintOptions: {},
  })

  expect(synthSnapshot(project)).toMatchSnapshot()
  expect(project.eslint).toBeInstanceOf(ESLint)
})

test('with Husky options', () => {
  const project = new Project({
    name: 'test-project',
    package: {},
    huskyOptions: {
      'pre-commit': 'lint-staged',
    },
  })

  expect(synthSnapshot(project)).toMatchSnapshot()
  expect(project.husky).toBeInstanceOf(Husky)
})

describe('with Jest configuration', () => {
  test('assigns jestConfig property', () => {
    const project = new Project({
      name: 'test-project',
      jestConfigOptions: {},
    })

    expect(project.jestConfig).toBeInstanceOf(JestConfig)
  })

  test('defaults', () => {
    const project = new Project({
      name: 'test-project',
      npmIgnoreOptions: {},
      jestConfigOptions: {},
    })

    expect(synthSnapshot(project)).toMatchSnapshot()
  })
})

describe('with license', () => {
  test('without author name', () => {
    expect(
      () =>
        new Project({
          name: 'test-project',
          package: {
            license: 'MIT',
          },
        }),
    ).toThrowError(/Missing package author name/)
  })

  test('assigns license property', () => {
    const project = new Project({
      name: 'test-project',
      package: {
        authorName: 'John Smith',
        license: 'MIT',
      },
    })

    expect(project.license).toBeInstanceOf(License)
  })

  test('with author name', () => {
    const project = new Project({
      name: 'test-project',
      package: {
        authorName: 'John Smith',
        license: 'MIT',
      },
    })

    expect(synthSnapshot(project)['license']).toMatchSnapshot()
  })

  test('with copyright year', () => {
    const project = new Project({
      name: 'test-project',
      package: {
        authorName: 'John Smith',
        copyrightYear: '2000',
        license: 'MIT',
      },
    })

    expect(synthSnapshot(project)['license']).toMatchSnapshot()
  })

  test('with author email', () => {
    const project = new Project({
      name: 'test-project',
      package: {
        authorEmail: 'john@example.com',
        authorName: 'John Smith',
        license: 'MIT',
      },
    })

    expect(synthSnapshot(project)['license']).toMatchSnapshot()
  })

  test('with author URL', () => {
    const project = new Project({
      name: 'test-project',
      package: {
        authorEmail: 'john@example.com',
        authorName: 'John Smith',
        authorUrl: 'https://example.com',
        license: 'MIT',
      },
    })

    expect(synthSnapshot(project)['license']).toMatchSnapshot()
  })
})

test('with `lint-staged`', () => {
  const project = new Project({
    name: 'test-project',
    lintStagedOptions: {},
  })

  expect(synthSnapshot(project)).toMatchSnapshot()
  expect(project.lintStaged).toBeInstanceOf(LintStaged)
})

describe('with NPM ignore', () => {
  test('assigns npmIgnore property', () => {
    const project = new Project({
      name: 'test-project',
      npmIgnoreOptions: {},
    })

    expect(project.npmIgnore).toBeInstanceOf(IgnoreFile)
  })

  test('defaults', () => {
    const project = new Project({
      name: 'test-project',
      npmIgnoreOptions: {},
    })

    expect(synthSnapshot(project)['.npmignore']).toMatchSnapshot()
  })
})

describe('with package', () => {
  test('assigns package property', () => {
    const project = new Project({
      name: 'test-project',
      package: {},
    })

    expect(project.package).toBeInstanceOf(NodePackage)
  })

  test('defaults', () => {
    const project = new Project({
      name: 'test-project',
      package: {},
    })

    expect(synthSnapshot(project)['package.json']).toMatchSnapshot()
  })
})

test('with Prettier options', () => {
  const project = new Project({
    name: 'test-project',
    prettierOptions: {},
  })

  expect(synthSnapshot(project)).toMatchSnapshot()
  expect(project.prettier).toBeInstanceOf(Prettier)
})

test('with Renovate options', () => {
  const project = new Project({
    name: 'test-project',
    renovateOptions: {},
  })

  expect(synthSnapshot(project)).toMatchSnapshot()
  expect(project.renovate).toBeInstanceOf(Renovate)
})

test('with Terraform enabled', () => {
  const project = new Project({
    name: 'test-project',
    withTerraform: true,
  })

  expect(synthSnapshot(project)).toMatchSnapshot()
})

test('with TypeScript options', () => {
  const project = new Project({
    name: 'test-project',
    package: {},
    typeScriptConfigOptions: {},
  })

  project.typeScriptConfig?.addFile('foo.js', 'bar.js')

  expect(synthSnapshot(project)).toMatchSnapshot()
  expect(project.typeScriptConfig).toBeInstanceOf(TypeScriptConfig)
})

test('with workspaces', () => {
  const project = new Project({
    name: 'test-project',
    eslintOptions: {},
    prettierOptions: {},
    workspaces: ['packages/*'],
  })

  expect(synthSnapshot(project)).toMatchSnapshot()
})
