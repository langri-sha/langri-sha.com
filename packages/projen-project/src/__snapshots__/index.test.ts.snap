// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaults 1`] = `
{
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
.*
!.babelrc
!.dockerignore
!.editorconfig
!.eslintignore
!.eslintrc.js
!.gitattributes
!.gitignore
!.gitkeep
!.husky
!.npmignore
!.openssl
!.prettierignore
*.db
*.log
!.github/
!.husky/
!.projen/
dist/
local/
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
",
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "files": [
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
}
`;

exports[`with Terraform enabled 1`] = `
{
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
.*
!.babelrc
!.dockerignore
!.editorconfig
!.eslintignore
!.eslintrc.js
!.gitattributes
!.gitignore
!.gitkeep
!.husky
!.npmignore
!.openssl
!.prettierignore
!.terraform.lock.hcl
*.db
*.log
!.github/
!.husky/
!.projen/
dist/
local/
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
",
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "files": [
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
}
`;

exports[`with TypeScript enabled 1`] = `
{
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
.*
!.babelrc
!.dockerignore
!.editorconfig
!.eslintignore
!.eslintrc.js
!.gitattributes
!.gitignore
!.gitkeep
!.husky
!.npmignore
!.openssl
!.prettierignore
*.db
*.log
*.tsbuildinfo
!.github/
!.husky/
!.projen/
dist/
local/
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
",
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "files": [
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
}
`;
